Android N FileUriExposedException:  exposed beyond app through ClipData.Item.getUri()

On Android N, When trying to capture image from camera the following ex raised: FileUriExposedException:  exposed beyond app through ClipData.Item.getUri()

```
     Caused by: java.lang.SecurityException: 
Permission Denial: opening provider android.support.v4.content.FileProvider from ProcessRecord{24ff435 15339:com.android.gallery3d/u0a44} (pid=15339, uid=10044) that is not exported from uid 10138
```

> StrictMode.VmPolicy.Builder builder = new StrictMode.VmPolicy.Builder();
> StrictMode.setVmPolicy(builder.build());
> 
> that is work for me.

     Caused by: java.lang.SecurityException: Permission Denial: opening provider android.support.v4.content.FileProvider from ProcessRecord{24ff435 15339:com.android.gallery3d/u0a44} (pid=15339, uid=10044) that is not exported from uid 10138


ava.lang.IllegalStateException: Binder has been finalized!
        at android.os.BinderProxy.transactNative(Native Method)
        at android.os.BinderProxy.transact(Binder.java:622)

--- a/.idea/gradle.xml
+++ b/.idea/gradle.xml
@@ -3,8 +3,10 @@
   <component name="GradleSettings">
     <option name="linkedExternalProjectsSettings">
       <GradleProjectSettings>
+        <option name="testRunner" value="PLATFORM" />
         <option name="distributionType" value="DEFAULT_WRAPPED" />
         <option name="externalProjectPath" value="$PROJECT_DIR$" />
+        <option name="gradleJvm" value="1.8" />
         <option name="modules">
           <set>
             <option value="$PROJECT_DIR$" />

--- a/.idea/misc.xml
+++ b/.idea/misc.xml
@@ -1,39 +1,15 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
+  <component name="CMakeSettings">
+    <configurations>
+      <configuration PROFILE_NAME="Debug" CONFIG_NAME="Debug" />
+    </configurations>
+  </component>
   <component name="MavenImportPreferences">
     <option name="generalSettings">
       <MavenGeneralSettings>
         <option name="mavenHome" value="Bundled (Maven 3)" />
       </MavenGeneralSettings>
-    </option>
-  </component>
-  <component name="NullableNotNullManager">
-    <option name="myDefaultNullable" value="android.support.annotation.Nullable" />
-    <option name="myDefaultNotNull" value="android.support.annotation.NonNull" />
-    <option name="myNullables">
-      <value>
-        <list size="7">
-          <item index="0" class="java.lang.String" itemvalue="org.jetbrains.annotations.Nullable" />
-          <item index="1" class="java.lang.String" itemvalue="javax.annotation.Nullable" />
-          <item index="2" class="java.lang.String" itemvalue="javax.annotation.CheckForNull" />
-          <item index="3" class="java.lang.String" itemvalue="edu.umd.cs.findbugs.annotations.Nullable" />
-          <item index="4" class="java.lang.String" itemvalue="android.support.annotation.Nullable" />
-          <item index="5" class="java.lang.String" itemvalue="androidx.annotation.Nullable" />
-          <item index="6" class="java.lang.String" itemvalue="androidx.annotation.RecentlyNullable" />
-        </list>
-      </value>
-    </option>
-    <option name="myNotNulls">
-      <value>
-        <list size="6">
-          <item index="0" class="java.lang.String" itemvalue="org.jetbrains.annotations.NotNull" />
-          <item index="1" class="java.lang.String" itemvalue="javax.annotation.Nonnull" />
-          <item index="2" class="java.lang.String" itemvalue="edu.umd.cs.findbugs.annotations.NonNull" />
-          <item index="3" class="java.lang.String" itemvalue="android.support.annotation.NonNull" />
-          <item index="4" class="java.lang.String" itemvalue="androidx.annotation.NonNull" />
-          <item index="5" class="java.lang.String" itemvalue="androidx.annotation.RecentlyNonNull" />
-        </list>
-      </value>
     </option>
   </component>
   <component name="ProjectRootManager" version="2" languageLevel="JDK_1_7" project-jdk-name="1.8" project-jdk-type="JavaSDK">

--- a/.idea/modules.xml
+++ b/.idea/modules.xml
@@ -2,9 +2,9 @@
 <project version="4">
   <component name="ProjectModuleManager">
     <modules>
-      <module fileurl="file://$PROJECT_DIR$/android-media-picker.iml" filepath="$PROJECT_DIR$/android-media-picker.iml" />
-      <module fileurl="file://$PROJECT_DIR$/app/app.iml" filepath="$PROJECT_DIR$/app/app.iml" />
-      <module fileurl="file://$PROJECT_DIR$/mediapicker/mediapicker.iml" filepath="$PROJECT_DIR$/mediapicker/mediapicker.iml" />
+      <module fileurl="file://$PROJECT_DIR$/.idea/modules/android-media-picker.iml" filepath="$PROJECT_DIR$/.idea/modules/android-media-picker.iml" group="android-media-picker" />
+      <module fileurl="file://$PROJECT_DIR$/.idea/modules/app/app.iml" filepath="$PROJECT_DIR$/.idea/modules/app/app.iml" group="android-media-picker/app" />
+      <module fileurl="file://$PROJECT_DIR$/.idea/modules/mediapicker/mediapicker.iml" filepath="$PROJECT_DIR$/.idea/modules/mediapicker/mediapicker.iml" group="android-media-picker/mediapicker" />
     </modules>
   </component>
 </project>

--- a/build.gradle
+++ b/build.gradle
@@ -9,10 +9,10 @@ buildscript {
         google()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.2.1'
+        classpath 'com.android.tools.build:gradle:4.0.1'
         classpath deps.kotlin.plugin
         classpath 'com.github.dcendents:android-maven-gradle-plugin:2.1'
-        classpath 'com.jfrog.bintray.gradle:gradle-bintray-plugin:1.8.0'
+        classpath 'com.jfrog.bintray.gradle:gradle-bintray-plugin:1.8.5'
         // NOTE: Do not place your application dependencies here; they belong
         // in the individual module build.gradle files
     }

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,6 @@
-#Tue Sep 04 17:12:37 ICT 2018
+#Tue Jan 15 13:47:56 ICT 2019
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.1.1-all.zip

--- a/mediapicker/src/main/AndroidManifest.xml
+++ b/mediapicker/src/main/AndroidManifest.xml
@@ -1,25 +1,27 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-          package="vn.tungdx.mediapicker">
+    package="vn.tungdx.mediapicker">
 
-    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
-    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
-    <uses-permission android:name="android.permission.CAMERA"/>
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.CAMERA" />
 
     <application>
         <activity
             android:name="vn.tungdx.mediapicker.activities.MediaPickerActivity"
             android:screenOrientation="portrait"
-            android:theme="@style/MediaPickerTheme"/>
+            android:theme="@style/MediaPickerTheme" />
 
         <provider
             android:name="androidx.core.content.FileProvider"
-            android:authorities="${applicationId}.fileprovider"
+            android:authorities="${applicationId}.provider"
             android:exported="false"
             android:grantUriPermissions="true">
+
             <meta-data
                 android:name="android.support.FILE_PROVIDER_PATHS"
-                android:resource="@xml/file_paths"/>
+                android:resource="@xml/file_paths" />
+
         </provider>
 
     </application>

--- a/mediapicker/src/main/java/vn/tungdx/mediapicker/MediaItem.kt
+++ b/mediapicker/src/main/java/vn/tungdx/mediapicker/MediaItem.kt
@@ -132,13 +132,12 @@ class MediaItem : Parcelable {
             return uri.path
         } else if (scheme == ContentResolver.SCHEME_CONTENT) {
             return if (isPhoto) {
-                MediaUtils.getRealImagePathFromURI(
-                        context.contentResolver, uri)
+                MediaUtils.getRealImagePathFromURI(context.contentResolver, uri)
             } else {
-                MediaUtils.getRealVideoPathFromURI(
-                        context.contentResolver, uri)
+                MediaUtils.getRealVideoPathFromURI(context.contentResolver, uri)
             }
         }
+
         return uri.toString()
     }
 

--- a/mediapicker/src/main/java/vn/tungdx/mediapicker/MediaOptions.kt
+++ b/mediapicker/src/main/java/vn/tungdx/mediapicker/MediaOptions.kt
@@ -6,7 +6,6 @@ import android.os.Parcelable
 import vn.tungdx.mediapicker.MediaOptions.Builder
 import vn.tungdx.mediapicker.utils.MediaUtils
 import java.io.File
-import java.util.*
 
 /**
  * @author TUNGDX
@@ -42,31 +41,42 @@ class MediaOptions : Parcelable {
     private var canSelectMultiVideo: Boolean = false
     var isCropped: Boolean = false
         private set
+
     /**
      *
      * @return in milliseconds.
      */
     var maxVideoDuration: Int = 0
         private set
+
     /**
      *
      * @return in milliseconds.
      */
     var minVideoDuration: Int = 0
         private set
+
     private var canSelectPhoto: Boolean = false
+
     private var canSelectVideo: Boolean = false
+
     var photoFile: File? = null
         private set
+
     var aspectX: Int = 0
         private set
+
     var aspectY: Int = 0
         private set
+
     var isFixAspectRatio: Boolean = false
         private set
+
     var croppedFile: File? = null
         private set
-    var mediaListSelected: List<MediaItem>? = ArrayList()
+
+    var mediaListSelected: List<MediaItem>? = null
+
     var isShowWarningVideoDuration: Boolean = false
         private set
 
@@ -149,7 +159,7 @@ class MediaOptions : Parcelable {
          * @param mediaSelecteds
          * Media list selected before.
          */
-        fun setMediaListSelected(mediaSelecteds: List<MediaItem>): Builder {
+        fun setMediaListSelected(mediaSelecteds: List<MediaItem>?): Builder {
             this.mediaListSelected = mediaSelecteds
             return this
         }
@@ -368,21 +378,21 @@ class MediaOptions : Parcelable {
         dest.writeTypedList(mediaListSelected)
     }
 
-    constructor(`in`: Parcel) {
-        canSelectMultiPhoto = `in`.readInt() != 0
-        canSelectMultiVideo = `in`.readInt() != 0
-        canSelectPhoto = `in`.readInt() != 0
-        canSelectVideo = `in`.readInt() != 0
-        isCropped = `in`.readInt() != 0
-        isFixAspectRatio = `in`.readInt() != 0
-        isShowWarningVideoDuration = `in`.readInt() != 0
-        this.maxVideoDuration = `in`.readInt()
-        this.minVideoDuration = `in`.readInt()
-        aspectX = `in`.readInt()
-        aspectY = `in`.readInt()
-        this.photoFile = `in`.readSerializable() as File?
-        this.croppedFile = `in`.readSerializable() as File?
-        `in`.readTypedList(this.mediaListSelected, MediaItem.CREATOR)
+    constructor(parcel: Parcel) {
+        canSelectMultiPhoto = parcel.readInt() != 0
+        canSelectMultiVideo = parcel.readInt() != 0
+        canSelectPhoto = parcel.readInt() != 0
+        canSelectVideo = parcel.readInt() != 0
+        isCropped = parcel.readInt() != 0
+        isFixAspectRatio = parcel.readInt() != 0
+        isShowWarningVideoDuration = parcel.readInt() != 0
+        this.maxVideoDuration = parcel.readInt()
+        this.minVideoDuration = parcel.readInt()
+        aspectX = parcel.readInt()
+        aspectY = parcel.readInt()
+        this.photoFile = parcel.readSerializable() as File?
+        this.croppedFile = parcel.readSerializable() as File?
+        parcel.readTypedList(this.mediaListSelected ?: listOf(), MediaItem.CREATOR)
     }
 
     override fun hashCode(): Int {

--- a/mediapicker/src/main/java/vn/tungdx/mediapicker/activities/MediaPickerActivity.kt
+++ b/mediapicker/src/main/java/vn/tungdx/mediapicker/activities/MediaPickerActivity.kt
@@ -28,6 +28,7 @@ import vn.tungdx.mediapicker.imageloader.MediaImageLoaderImpl
 import vn.tungdx.mediapicker.utils.MediaUtils
 import vn.tungdx.mediapicker.utils.MessageUtils
 import vn.tungdx.mediapicker.utils.RecursiveFileObserver
+import vn.tungdx.mediapicker.utils.Utils
 import java.io.File
 import java.io.IOException
 import java.util.*
@@ -248,7 +249,7 @@ class MediaPickerActivity : AppCompatActivity(), MediaSelectedListener, CropList
             var file: File? = mMediaOptions!!.photoFile
             if (file == null) {
                 try {
-                    file = MediaUtils.createDefaultImageFile()
+                    file = MediaUtils.createDefaultImageFile(this)
                 } catch (e: IOException) {
                     e.printStackTrace()
                 }
@@ -256,8 +257,7 @@ class MediaPickerActivity : AppCompatActivity(), MediaSelectedListener, CropList
             }
             if (file != null) {
                 mPhotoFileCapture = file
-                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT,
-                        Uri.fromFile(file))
+                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, Utils.getUriForFile(this, file))
                 startActivityForResult(takePictureIntent, REQUEST_PHOTO_CAPTURE)
                 mFileObserverTask = FileObserverTask()
                 mFileObserverTask!!.execute()
@@ -425,8 +425,7 @@ class MediaPickerActivity : AppCompatActivity(), MediaSelectedListener, CropList
             // not get duration by Uri after record video.(It's usually happen
             // in HTC
             // devices 2.3, maybe others)
-            duration = MediaUtils
-                    .getDuration(applicationContext, videoUri)
+            duration = MediaUtils.getDuration(applicationContext, videoUri)
         }
         // accept delta about < 1000 milliseconds. (ex: 10769 is still accepted
         // if limit is 10000)
@@ -540,7 +539,7 @@ class MediaPickerActivity : AppCompatActivity(), MediaSelectedListener, CropList
                                             grantResults: IntArray) {
         when (requestCode) {
             REQUEST_CAMERA_PERMISSION -> {
-                if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                     if (takePhotoPending) {
                         takePhoto()
                     } else if (takeVideoPending) {
@@ -553,10 +552,8 @@ class MediaPickerActivity : AppCompatActivity(), MediaSelectedListener, CropList
         }
         //pass permission result to fragments in this activity
         val fragments = supportFragmentManager.fragments
-        if (fragments != null) {
-            for (fragment in fragments) {
-                fragment.onRequestPermissionsResult(requestCode, permissions, grantResults)
-            }
+        for (fragment in fragments) {
+            fragment.onRequestPermissionsResult(requestCode, permissions, grantResults)
         }
     }
 
@@ -564,6 +561,7 @@ class MediaPickerActivity : AppCompatActivity(), MediaSelectedListener, CropList
         private const val TAG = "MediaPickerActivity"
 
         const val EXTRA_MEDIA_OPTIONS = "extra_media_options"
+
         /**
          * Intent extra included when return back data in
          * [Activity.onActivityResult] of activity or fragment

--- a/mediapicker/src/main/java/vn/tungdx/mediapicker/utils/MediaUtils.kt
+++ b/mediapicker/src/main/java/vn/tungdx/mediapicker/utils/MediaUtils.kt
@@ -51,12 +51,11 @@ object MediaUtils {
      * @throws IOException
      */
     @Throws(IOException::class)
-    fun createDefaultImageFile(): File {
+    fun createDefaultImageFile(context: Context): File {
         val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss",
                 Locale.getDefault()).format(Date())
         val imageFileName = "JPEG_$timeStamp.jpg"
-        val storageDir = Environment
-                .getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
+        val storageDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES)!!
         if (!storageDir.exists()) {
             storageDir.mkdirs()
         }
@@ -176,17 +175,18 @@ object MediaUtils {
     fun getRealVideoPathFromURI(contentResolver: ContentResolver,
                                 contentURI: Uri): String? {
         val cursor = contentResolver.query(contentURI, null, null, null, null)
-        if (cursor == null)
-            return contentURI.path
-        else {
-            cursor.moveToFirst()
-            val idx = cursor.getColumnIndex(Video.VideoColumns.DATA)
-            try {
-                return cursor.getString(idx)
-            } catch (exception: Exception) {
-                return null
-            }
+                ?: return contentURI.path
 
+        cursor.moveToFirst()
+
+        val idx = cursor.getColumnIndex(Video.VideoColumns.DATA)
+
+        return try {
+            val path = cursor.getString(idx)
+            cursor.close()
+            path
+        } catch (exception: Exception) {
+            null
         }
     }
 

--- a/mediapicker/src/main/java/vn/tungdx/mediapicker/utils/Utils.kt
+++ b/mediapicker/src/main/java/vn/tungdx/mediapicker/utils/Utils.kt
@@ -3,14 +3,13 @@ package vn.tungdx.mediapicker.utils
 import android.annotation.TargetApi
 import android.app.Activity
 import android.content.Context
-import android.content.res.TypedArray
+import android.net.Uri
 import android.os.Build
 import android.os.Environment
-
+import androidx.core.content.FileProvider
+import vn.tungdx.mediapicker.R
 import java.io.File
 import java.io.IOException
-
-import vn.tungdx.mediapicker.R
 
 /**
  * Created by TungDX on 6/4/2015.
@@ -71,5 +70,10 @@ object Utils {
     fun createTempFile(context: Context, folder: File?): File {
         val prefix = System.currentTimeMillis().toString()
         return File.createTempFile(prefix, null, folder)
+    }
+
+    @JvmStatic
+    fun getUriForFile(context: Context, file: File): Uri {
+        return FileProvider.getUriForFile(context, context.applicationContext.packageName + ".provider", file)
     }
 }

--- a/mediapicker/src/main/res/xml/file_paths.xml
+++ b/mediapicker/src/main/res/xml/file_paths.xml
@@ -2,5 +2,8 @@
 <paths>
     <external-path
         name="images"
-        path="Pictures"/>
+        path="Pictures" />
+    <external-files-path
+        name="external_images"
+        path="Pictures" />
 </paths>

--- a/versions.gradle
+++ b/versions.gradle
@@ -7,7 +7,7 @@ def deps = [:]
 
 def versions = [
         androidx: "1.0.0",
-        kotlin  : "1.2.70",
+        kotlin  : "1.4.0",
         glide   : "4.8.0"
 ]
 
@@ -34,9 +34,9 @@ ext.deps = deps
 
 def build_versions = [
         min_sdk    : 15,
-        target_sdk : 28,
-        compile_sdk: 28,
-        build_tools: "28.0.3"
+        target_sdk : 30,
+        compile_sdk: 30,
+        build_tools: "30.0.2"
 ]
 ext.build_versions = build_versions
 

